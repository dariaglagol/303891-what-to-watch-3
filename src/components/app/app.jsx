import React from "react";
import PropTypes from "prop-types";
import {BrowserRouter, Switch, Route} from "react-router-dom";
import {connect} from "react-redux";

import Main from "@components/main/main";
import MovieExtended from "@components/movie-extended/movie-extended";
import withTabs from "@hocs/with-tabs/with-tabs";
import withCatalog from "@hocs/with-catalog/with-catalog";
import withMovieList from "@hocs/with-movie-list/with-movie-list";
import withReview from "@hocs/with-review/with-review";
import {PageTypes} from "@utils/constants";

import {ActionCreator as CommonActionCreator} from "@reducers/common/common";
import {getActivePage, getFullScreenPlayerState} from "@reducers/common/selectors";

import {Operation as UserOperation} from "@reducers/user/user";
import {getAuthStatus, getUserData, getUserError} from "@reducers/user/selectors";

import {ActionCreator as DataActionCreator, Operation as DataOperation} from "@reducers/data/data";
import {
  getActiveGenre,
  getFilmsSelector,
  getReviews,
  getMovieCover,
  getActiveFilmId,
  getReviewError
} from "@reducers/data/selectors.js";

import Loading from "@components/loading/loading";
import SignIn from "@components/sign-in/sign-in";
import AddReview from "@components/add-review/add-review";

const MovieExtendedComponentWrapped = withMovieList(withTabs(MovieExtended));
const MainComponentWrapped = withMovieList(withCatalog(Main));
const ReviewComponentWrapped = withReview(AddReview);

const App = (props) => {
  const {
    promoMovie,
    films,
    reviews,
    activeGenre,
    onGenreTabClick,
    activePage,
    onPageChange,
    isFullscreenPlayerActive,
    onFullScreenToggle,
    activeFilmId,
    login,
    userData,
    userErrors,
    authStatus,
    addReview,
    reviewError,
  } = props;

  const movieDetails = films.find((film) => film.id === activeFilmId) || {};

  function _renderPages() {
    switch (activePage) {
      case PageTypes.MAIN:
        return (
          <MainComponentWrapped
            userData={userData}
            authStatus={authStatus}
            promoMovie={promoMovie}
            onFilmClick={onPageChange}
            onSignInClick={onPageChange}
            activeGenre={activeGenre}
            onGenreTabClick={onGenreTabClick}
            isFullscreenPlayerActive={isFullscreenPlayerActive}
            onFullScreenToggle={onFullScreenToggle}
            films={films}
          />
        );
      case PageTypes.MOVIE:
        return (
          <MovieExtendedComponentWrapped
            userData={userData}
            authStatus={authStatus}
            onAddReviewClick={onPageChange}
            onFilmClick={onPageChange}
            onSignInClick={onPageChange}
            films={films}
            movieDetails={movieDetails}
            isFullscreenPlayerActive={isFullscreenPlayerActive}
            onFullScreenToggle={onFullScreenToggle}
            reviews={reviews}
          />
        );
      case PageTypes.LOADING:
        return (
          <Loading />
        );
      case PageTypes.AUTH:
        return (
          <SignIn
            onSubmit={login}
            userErrors={userErrors}
          />
        );
      case PageTypes.REVIEW:
        return (
          <ReviewComponentWrapped
            userData={userData}
            authStatus={authStatus}
            movieDetails={movieDetails}
            onSignInClick={onPageChange}
            onSubmit={addReview}
            reviewError={reviewError}
          />
        );
    }

    return null;
  }

  return (
    <BrowserRouter>
      <Switch>
        <Route exact path="/">
          {_renderPages()}
        </Route>
        <Route exact path="/dev-movie-details">
          <MovieExtendedComponentWrapped
            userData={userData}
            authStatus={authStatus}
            onFilmClick={onPageChange}
            onSignInClick={onPageChange}
            onAddReviewClick={onPageChange}
            films={films}
            movieDetails={movieDetails}
            isFullscreenPlayerActive={isFullscreenPlayerActive}
            onFullScreenToggle={onFullScreenToggle}
            reviews={reviews}
          />
        </Route>
        <Route exact path="/dev-review">
          <ReviewComponentWrapped
            userData={userData}
            authStatus={authStatus}
            movieDetails={movieDetails}
            onSignInClick={onPageChange}
            onSubmit={addReview}
            reviewError={reviewError}
          />
        </Route>
      </Switch>
    </BrowserRouter>
  );
};

App.defaultProps = {
  films: [],
  promoMovie: {},
  reviews: [],
  movieDetails: {},
  userData: {},
};

const mapStateToProps = (state) => ({
  films: getFilmsSelector(state),
  filteredFilms: getFilmsSelector(state),
  activeGenre: getActiveGenre(state),
  promoMovie: getMovieCover(state),
  reviews: getReviews(state),
  activePage: getActivePage(state),
  activeFilmId: getActiveFilmId(state),
  isFullscreenPlayerActive: getFullScreenPlayerState(state),
  authStatus: getAuthStatus(state),
  userData: getUserData(state),
  userErrors: getUserError(state),
  reviewError: getReviewError(state),
});

const mapDispatchToProps = (dispatch) => ({
  onGenreTabClick(activeGenre) {
    dispatch(DataActionCreator.changeGenre(activeGenre));
  },
  onPageChange(activePage, id) {
    if (id) {
      dispatch(DataActionCreator.getActiveFilmId(id));
    }
    dispatch(CommonActionCreator.setActivePage(activePage));
  },
  onFullScreenToggle(state) {
    dispatch(CommonActionCreator.toggleFullscreenPlayer(state));
  },
  login(authData) {
    dispatch(UserOperation.login(authData));
  },
  addReview(reviewData) {
    dispatch(DataOperation.sendReview(reviewData));
  }
});

App.propTypes = {
  promoMovie: PropTypes.oneOfType([
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      posterImage: PropTypes.string.isRequired,
      previewImage: PropTypes.string.isRequired,
      backgroundImage: PropTypes.string.isRequired,
      backgroundColor: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      rating: PropTypes.number.isRequired,
      scoresCount: PropTypes.number.isRequired,
      director: PropTypes.string.isRequired,
      starring: PropTypes.array.isRequired,
      runTime: PropTypes.number.isRequired,
      genre: PropTypes.string.isRequired,
      released: PropTypes.number.isRequired,
      id: PropTypes.number.isRequired,
      isFavorite: PropTypes.bool.isRequired,
      videoLink: PropTypes.string.isRequired,
      previewVideoLink: PropTypes.string.isRequired,
    }),
    PropTypes.exact({}).isRequired
  ]).isRequired,
  films: PropTypes.oneOfType([
    PropTypes.exact([]).isRequired,
    PropTypes.arrayOf(PropTypes.exact({
      name: PropTypes.string.isRequired,
      posterImage: PropTypes.string.isRequired,
      previewImage: PropTypes.string.isRequired,
      backgroundImage: PropTypes.string.isRequired,
      backgroundColor: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      rating: PropTypes.number.isRequired,
      scoresCount: PropTypes.number.isRequired,
      director: PropTypes.string.isRequired,
      starring: PropTypes.array.isRequired,
      runTime: PropTypes.number.isRequired,
      genre: PropTypes.string.isRequired,
      released: PropTypes.number.isRequired,
      id: PropTypes.number.isRequired,
      isFavorite: PropTypes.bool.isRequired,
      videoLink: PropTypes.string.isRequired,
      previewVideoLink: PropTypes.string.isRequired,
    })).isRequired,
  ]).isRequired,
  reviews: PropTypes.arrayOf(PropTypes.exact({
    text: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
  })),
  activeGenre: PropTypes.exact({
    multiply: PropTypes.string.isRequired,
    single: PropTypes.string.isRequired,
  }).isRequired,
  onGenreTabClick: PropTypes.func.isRequired,
  onPageChange: PropTypes.func.isRequired,
  activePage: PropTypes.string.isRequired,
  isFullscreenPlayerActive: PropTypes.bool.isRequired,
  onFullScreenToggle: PropTypes.func.isRequired,
  activeFilmId: PropTypes.number.isRequired,
  authStatus: PropTypes.string.isRequired,
  userData: PropTypes.oneOfType([
    PropTypes.exact({
      id: PropTypes.number.isRequired,
      email: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      avatarUrl: PropTypes.string.isRequired,
    }),
    PropTypes.exact({})
  ]).isRequired,
  login: PropTypes.func.isRequired,
  userErrors: PropTypes.shape({
    error: PropTypes.string,
  })
};

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);


