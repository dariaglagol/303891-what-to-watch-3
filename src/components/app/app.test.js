import React from "react";
import renderer from "react-test-renderer";
import {App} from "./app";

const Mocks = {
  mockFilmData: {
    name: `The Grand Budapest Hotel`,
    genre: `Comedy`,
    released: `2020`,
    poster: `img/bohemian-rhapsody.jpg`,
    previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
  },
  mockCatalogFilms: [
    {
      name: `Fantastic Beasts: The Crimes of Grindelwald`,
      poster: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
      genre: `Comedy`,
      previewVideoLink: `previewVideoLink`,

    },
    {
      name: `Bohemian Rhapsody`,
      poster: `img/bohemian-rhapsody.jpg`,
      genre: `Horror`,
      previewVideoLink: `previewVideoLink`,
    },
    {
      name: `Macbeth`,
      poster: `img/macbeth.jpg`,
      genre: `Drama`,
      previewVideoLink: `previewVideoLink`,
    },
    {
      name: `Aviator`,
      poster: `img/aviator.jpg`,
      genre: `Comedy`,
      previewVideoLink: `previewVideoLink`,
    },
    {
      name: `We need to talk about Kevin`,
      poster: `img/we-need-to-talk-about-kevin.jpg`,
      genre: `Science`,
      previewVideoLink: `previewVideoLink`,
    },
    {
      name: `What We Do in the Shadows`,
      poster: `img/what-we-do-in-the-shadows.jpg`,
      genre: `Detective`,
      previewVideoLink: `previewVideoLink`,
    },
    {
      name: `Revenant`,
      poster: `img/revenant.jpg`,
      genre: `Thriller`,
      previewVideoLink: `previewVideoLink`,
    },
    {
      name: `Johnny English`,
      poster: `img/johnny-english.jpg`,
      genre: `Historical`,
      previewVideoLink: `previewVideoLink`,
    }
  ],
  mockMovieDetails: {
    name: `The Grand Budapest Hotel`,
    genre: `Comedy`,
    poster: `img/bg-the-grand-budapest-hotel.jpg`,
    released: `2020`,
    score: 8.9,
    rating: 124,
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,
    director: `Wes Anderson`,
    starring: `Bill Murray, Edward Norton, Jude Law, Willem Dafoe and other`,
    runTime: 113,
    previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  },
  defaultActiveGenre: {
    single: `All genres`,
    multiply: `All genres`
  },
  defaultActivePage: `main`,
  isFullscreenPlayerActive: false,
};

it(`Render App`, () => {
  const {mockFilmData, mockCatalogFilms, mockMovieDetails, defaultActiveGenre, defaultActivePage, isFullscreenPlayerActive} = Mocks;
  const appComponent = renderer
    .create(
        <App
          promoMovie={mockFilmData}
          films={mockCatalogFilms}
          movieDetails={mockMovieDetails}
          activeGenre={defaultActiveGenre}
          onGenreTabClick={() => {}}
          onPageChange={() => {}}
          onFullScreenToggle={() => {}}
          activePage={defaultActivePage}
          isFullscreenPlayerActive={isFullscreenPlayerActive}
        />
        , {createNodeMock: () => {
          return {};
        }}
    )
    .toJSON();

  expect(appComponent).toMatchSnapshot();
});
