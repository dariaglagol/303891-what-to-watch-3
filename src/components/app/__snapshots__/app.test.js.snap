// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render App 1`] = `
Array [
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="name"
        src="backgroundImage"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <a
          className="user-block__link"
          href="/login"
          onClick={[Function]}
        >
          Sign in
        </a>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <a
            href="/films/1"
            onClick={[Function]}
          >
            <img
              alt="The Grand Budapest Hotel poster"
              height="327"
              src="posterImage"
              width="218"
            />
          </a>
        </div>
        <div
          className="movie-card__desc"
        >
          <a
            href="/films/1"
            onClick={[Function]}
          >
            <h2
              className="movie-card__title"
            >
              name
            </h2>
          </a>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              genre
            </span>
            <span
              className="movie-card__year"
            >
              2020
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <a
              href="/player/1"
              onClick={[Function]}
            >
              <button
                className="btn btn--play movie-card__button"
                type="button"
              >
                <svg
                  height="19"
                  viewBox="0 0 19 19"
                  width="19"
                >
                  <use
                    xlinkHref="#play-s"
                  />
                </svg>
                <span>
                  Play
                </span>
              </button>
            </a>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
         Catalog 
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item catalog__genres-item--active"
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            All genres
          </a>
        </li>
        <li
          className="catalog__genres-item "
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Comedy
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            href="/films/1"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                className=""
                poster="backgroundImage"
              >
                <source
                  src="previewVideoLink"
                />
              </video>
            </div>
            <h3
              className="small-movie-card__title"
            >
              <span
                className="small-movie-card__link"
              >
                name
              </span>
            </h3>
          </a>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            href="/films/2"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                className=""
                poster="backgroundImage 2"
              >
                <source
                  src="previewVideoLink 2"
                />
              </video>
            </div>
            <h3
              className="small-movie-card__title"
            >
              <span
                className="small-movie-card__link"
              >
                name 2
              </span>
            </h3>
          </a>
        </article>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
